src/Native/Unix/System.Native/pal_networking.c

```diff
@@ -1650,12 +1651,35 @@ int32_t SystemNative_Accept(intptr_t socket, uint8_t* socketAddress, int32_t* so
     while ((accepted = accept4(fd, (struct sockaddr*)socketAddress, &addrLen, SOCK_CLOEXEC)) < 0 && errno == EINTR);
 #else
     while ((accepted = accept(fd, (struct sockaddr*)socketAddress, &addrLen)) < 0 && errno == EINTR);
+#if defined(FD_CLOEXEC)
+    // macOS does not have accept4 but it can set _CLOEXEC on descriptor.
+    if (accepted != -1)
+    {
+        if (fcntl(accepted, F_SETFD, FD_CLOEXEC) != 0)
+        {
+            int oldError = errno;
+            close(accepted);
+            *acceptedSocket = -1;
+            return SystemNative_ConvertErrorPlatformToPal(oldError);
+        }
+    }
+#endif
 #endif
     if (accepted == -1)
     {
         *acceptedSocket = -1;
         return SystemNative_ConvertErrorPlatformToPal(errno);
     }
+#if !defined(__linux__)
+    // On macOS and FreeBSD new socket inherits flags from accepting fd.
+    if (SystemNative_FcntlSetIsNonBlocking(accepted, 0) != 0)
+    {
+        int oldError = errno;
+        close(accepted);
+        *acceptedSocket = -1;
+        return SystemNative_ConvertErrorPlatformToPal(oldError);
+    }
+#endif
```

Comment:
Similarly we could move this block up to above the -1 check above, delete this return line, and just have all of the -1 checking and error code handling consolidated.